### DISCUSSION TITLE
The three types of address maps

### DISCUSSION THREAD

**Discussion by RasmusGOlsen** [SystemRDL/systemrdl-compiler#91] [General]
If I categorize the use of `addrmap` in SystemRDL into three different types, I get:

1. A leaf map only containing registers, i.e. `regfile` and `reg`.
2. A hierarchical map only containing other `addrmap`s.
3. A combination of 1 and 2 containing both registers and address maps.

Is the third option a good idea? Does it make sense to generate RTL code from options 2 and 3 then how?

**Comment by amykyta3** [comment]
For 2, the addrmap can be interpreted as an address decode stage to multiple sub-addrmap register blocks.
So for example, you could make an RTL generator that creates an AX4-Lite component that decodes 1-to-N devices. A slave port transaction would be dispatched to the appropriate master port based on the address offsets of each child addrmap.

Agree that the 3rd situation makes less practical sense. I suppose you could have an RTL generator that creates a register, and simultaneously creates a bus interface port to an external addrmap? Seems unnecessarily convoluted though.

The thing to remember with SystemRDL is that it aims to describe the hardware behavior. The language makes very little attempt to describe the specifics of the RTL implementation. So situations like this are very much open to interpretation.
